File: src/main/java/cz/wa/secretagent/game/action/AgentIslandAction.java
//package cz.wa.secretagent.game.action;
//
//import org.apache.commons.lang.Validate;
//import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;
//
//import cz.wa.secretagent.world.entity.Entity;
//import cz.wa.secretagent.world.entity.agent.HumanAgent;
//import cz.wa.secretagent.world.entity.direction.EntityXDirection;
//import cz.wa.secretagent.world.entity.direction.EntityYDirection;
//import secretAgent.game.action.AgentAction;
//import secretAgent.game.action.AgentActivateAction;
//
///**
// * Action to control an agent on island map.
// *
// * @author Ondrej Milenovsky
// */
//public class AgentIslandAction extends AgentAction<HumanAgent> {
//
//    public AgentActivateAction activateAction;
//
//    @Override
//    public void init() {
//        activateAction = getActionFactory().getAction(AgentActivateAction.class);
//    }
//
//    public void moveX(EntityXDirection dir) {
//        Validate.notNull(dir, "dir is null");
//        HumanAgent agent = getAgent();
//        if (agent.isControlable()) {
//            double maxSpeed = agent.getCapabilities().getMaxSpeed();
//            Vector2D addSpeed = dir.getVector().scalarMultiply(maxSpeed);
//            agent.setMoveSpeed(new Vector2D(0, agent.getMoveSpeed().getY()).add(addSpeed));
//            if (dir != EntityXDirection.NONE) {
//                agent.setDirection(dir);
//            }
//        }
//    }
//
//    public void moveY(EntityYDirection dir) {
//        Validate.notNull(dir, "dir is null");
//        HumanAgent agent = getAgent();
//        if (agent.isControlable()) {
//            double maxSpeed = agent.getCapabilities().getMaxSpeed();
//            agent.setMoveSpeed(new Vector2D(agent.getMoveSpeed().getX(), 0).add(dir.getVector()
//                    .scalarMultiply(maxSpeed)));
//        }
//    }
//
//    public void activate(boolean b) {
//        if (!b) {
//            activateAction.stopUsing();
//            return;
//        }
//        HumanAgent agent = getAgent();
//        if (agent.isControlable() && agent.getCapabilities().canActivate()) {
//            Entity usable = agent.getEntityToUse();
//            if (usable != null) {
//                activateAction.useEntity(usable);
//            }
//        }
//    }
//
//}

File: src/main/kotlin/secretAgent/game/action/AgentIslandAction.kt
package secretAgent.game.action

import org.apache.commons.math3.geometry.euclidean.twod.Vector2D
import secretAgent.plus
import secretAgent.world.entity.EntityXDirection
import secretAgent.world.entity.EntityYDirection
import secretAgent.world.entity.HumanAgent

/**
 * Action to control an agent on island map.
 *
 * @author Ondrej Milenovsky
 */
class AgentIslandAction : AgentAction<HumanAgent>() {

    lateinit var activateAction: AgentActivateAction

    override fun init() {
        activateAction = actionFactory!!.getAction(AgentActivateAction::class.java)
    }

    fun moveX(dir: EntityXDirection) {
        val agent = agent!!
        if (agent.isControllable) {
            val maxSpeed = agent.capabilities.maxSpeed
            val addSpeed = dir * maxSpeed
            agent.moveSpeed = Vector2D(0.0, agent.moveSpeed.y) + addSpeed
            if (dir != EntityXDirection.NONE)
                agent.direction = dir
        }
    }

    fun moveY(dir: EntityYDirection) {
        val agent = agent!!
        if (agent.isControllable) {
            val maxSpeed = agent.capabilities.maxSpeed
            agent.moveSpeed = Vector2D(agent.moveSpeed.x, 0.0) + dir * maxSpeed
        }
    }

    fun activate(b: Boolean) {
        if (!b) {
            activateAction.stopUsing()
            return
        }
        val agent = agent!!
        if (agent.isControllable && agent.capabilities.canActivate)
            agent.entityToUse?.let(activateAction::useEntity)
    }
}